### DISCUSSION TITLE
Bit scanning operations in emscripten.

### DISCUSSION THREAD

**Discussion by ovostrikov** [emscripten-core/emscripten#17632] [Q&A] [answered]
I've been trying to port some code that relies on [vectorclass](https://github.com/vectorclass/version2) library to emscripten. While patching in proper includes and removing dynamic dispatching logic fixed most of compilation errors, there is one blocker remaining: bit scan forward/reverse functions.
Here's [link to corresponding code](https://github.com/vectorclass/version2/blob/master/instrset.h#:~:text=%23endif-,//%20Define%20bit%2Dscan%2Dforward%20function.%20Gives%20index%20to%20lowest%20set,%23endif,-//%20Same%20function).
It uses [built-in functions](https://docs.microsoft.com/en-us/cpp/intrinsics/bitscanforward-bitscanforward64?view=msvc-170) from <intrin.h> for msvc/intel and assembly snippet for gcc/clang. Neither works in emscripten for obvious reasons.

So are there any built-in analogues in emscripten compiler or should I fall back to less efficient C++ implementation?

**Comment by tlively** [comment, marked as answer]
There's no single instruction that has the same behavior, but you can use `__builtin_ctz` and `__builtin_popcount` to emulate it and those should lower to a single Wasm instruction each. Emscripten is a thin wrapper around clang at compile time, so it supports everything normal clang supports.
